src/CoreWCF.NetTcp/tests/ServiceBehaviorUseSynchronizationContextAsyncTests.cs

```diff
@@ -0,0 +1,169 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.ServiceModel.Channels;
+using System.Threading;
+using System.Threading.Tasks;
+using Contract;
+using CoreWCF.Configuration;
+using Helpers;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Services;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace CoreWCF.NetTcp.Tests
+{
+    /// <summary>
+    /// Verifies the effect of <see cref="ServiceBehaviorAttribute.UseSynchronizationContext"/> on asynchronous operations.
+    /// </summary>
+    public class ServiceBehaviorUseSynchronizationContextAsyncTests
+    {
+        public const string RelativeAddress = "/nettcp.svc/UseSynchronizationContextAsync";
+        private readonly ITestOutputHelper _output;
+
+        public ServiceBehaviorUseSynchronizationContextAsyncTests(ITestOutputHelper output)
+        {
+            _output = output;
+
+            TaskService.s_synchronizationContextWhenCreated = null;
+            TaskService.s_synchronizationContextWhenCalled = null;
```

Comment:
Passing data around via statics isn't a good idea as you can have problems with concurrent test execution. You could solve this by using a singleton in DI.  
1. Change the method signature of RunAsync to `private async Task<TService> RunAsync<TService>()`  
2. Add a singleton TService to DI in `Startup<TService>.ConfigureServices` : `services.AddSingleton<TService>();`  
3. Request the TService instance from DI in RunAsync after the test has run : `TService serviceInstance = host.Services.GetRequiredService<TService>();` and return it as the last statement in the using block.  

Other than the use of statics, the tests in this class look pretty thorough.

Reply:
Yes, I see what you mean. Didn't think of that. But this would also mean that the service implementations would need to access the singleton and that would require registering the singleton in all tests that use EchoService and ChatService, right? These two existed before and I just updated them a bit to store the context and thread id to make it possible to add the asserts in the new tests. I would prefer to add dedicated service implementations just for these tests, to not clutter all other tests with irrelevant registrations. Does that feel right?

Reply:
It has a default constructor so the non-DI mechanism will still work elsewhere. WCF on its own can instantiate a type with a default constructor with no DI injection. You would only break other tests if you added a required construction parameter. But I do agree with adding a custom type just for this test aa it's cleaner.